File: fastApi-agent-service/backend/utils/health_safety.py

from typing import Dict, List, Any, Optional
from backend.constants.enums import (
    ActivityLevel, Goal, MIN_CALORIES_ADULT, MAX_CALORIES_ADULT,
    MIN_WORKOUT_MINUTES, MAX_WORKOUT_MINUTES
)
import logging

logger = logging.getLogger(__name__)

class HealthSafetyValidator:
    """Validates health recommendations for safety"""
    
    @staticmethod
    def validate_calorie_target(calories: int, age: Optional[int] = None, goal: Goal = Goal.GENERAL_WELLNESS) -> Dict[str, Any]:
        """
        Validates calorie targets are within safe ranges
        Returns validation result and adjusted calories if needed
        """
        result = {
            "is_valid": True,
            "adjusted_calories": calories,
            "warnings": [],
            "recommendations": []
        }
        
        
        if calories < MIN_CALORIES_ADULT:
            result["is_valid"] = False
            result["adjusted_calories"] = MIN_CALORIES_ADULT
            result["warnings"].append(f"Calorie target too low. Adjusted to safe minimum: {MIN_CALORIES_ADULT}")
            
    
        if calories > MAX_CALORIES_ADULT:
            result["is_valid"] = False
            result["adjusted_calories"] = MAX_CALORIES_ADULT  
            result["warnings"].append(f"Calorie target too high. Adjusted to conservative maximum: {MAX_CALORIES_ADULT}")
            
     
        if goal == Goal.GENTLE_WEIGHT_LOSS and calories < 1500:
            result["warnings"].append("Very low calorie diets should only be undertaken with medical supervision")
            result["recommendations"].append("Consider consulting with a registered dietitian for personalized guidance")
            
        return result
    
    @staticmethod
    def validate_workout_plan(workout_minutes_per_day: int, activity_level: ActivityLevel, age: Optional[int] = None) -> Dict[str, Any]:
        """
        Validates workout duration and intensity for safety
        """
        result = {
            "is_valid": True,
            "adjusted_minutes": workout_minutes_per_day,
            "warnings": [],
            "recommendations": []
        }
        
     
        if workout_minutes_per_day < MIN_WORKOUT_MINUTES:
            result["adjusted_minutes"] = MIN_WORKOUT_MINUTES
            result["recommendations"].append("Even short workouts can be beneficial - aim for at least 10 minutes")
            
        if workout_minutes_per_day > MAX_WORKOUT_MINUTES:
            result["is_valid"] = False
            result["adjusted_minutes"] = MAX_WORKOUT_MINUTES
            result["warnings"].append("Workout duration exceeds recommended limits for general population")
            
        
        if activity_level == ActivityLevel.SEDENTARY and workout_minutes_per_day > 45:
            result["warnings"].append("Gradual increase in activity is recommended for sedentary individuals")
            result["recommendations"].append("Start with shorter workouts and build up gradually")
            
       
        if age and age > 65 and workout_minutes_per_day > 60:
            result["recommendations"].append("Adults over 65 should consult healthcare providers for exercise guidance")
            
        if age and age < 18:
            result["recommendations"].append("Youth fitness programs should be age-appropriate and supervised")
            
        return result
    
    @staticmethod
    def validate_dietary_restrictions(restrictions: List[str], goal: Goal) -> Dict[str, Any]:
        """
        Ensures dietary restrictions are compatible with health goals
        """
        result = {
            "is_valid": True,
            "warnings": [],
            "recommendations": []
        }
        
        
        if len(restrictions) > 3:
            result["warnings"].append("Multiple dietary restrictions may require professional nutrition guidance")
            result["recommendations"].append("Consider consulting with a registered dietitian")
            
        
        if "vegan" in restrictions and goal == Goal.MUSCLE_GAIN:
            result["recommendations"].append("Vegan muscle gain requires careful protein planning - consider plant-based protein sources")
            
        if "diabetic_friendly" in restrictions:
            result["recommendations"].append("Diabetic meal plans should be reviewed with healthcare provider")
            result["warnings"].append("Monitor blood sugar levels when following new meal plans")
            
        return result
    
    @staticmethod
    def get_safe_recommendation_disclaimer() -> str:
        """Returns standard disclaimer for all health recommendations"""
        return """
IMPORTANT: This is a general wellness suggestion and not medical advice. 
Please consult with qualified healthcare professionals before making significant 
changes to your diet or exercise routine, especially if you have health conditions, 
take medications, or are pregnant. Individual needs vary significantly, and 
personalized professional guidance is recommended for optimal health outcomes.
"""
    
    @staticmethod
    def validate_user_profile_safety(profile_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Validates user profile data for safety red flags
        """
        result = {
            "is_safe": True,
            "concerns": [],
            "recommendations": []
        }
        
        age = profile_data.get("age")
        goal = profile_data.get("primary_goal")
        health_conditions = profile_data.get("health_conditions", [])
        
        
        if age and (age < 16 or age > 75):
            result["concerns"].append("Age requires special consideration for health planning")
            result["recommendations"].append("Professional supervision recommended")
            
       
        high_risk_conditions = [
            "heart disease", "diabetes", "high blood pressure", "eating disorder",
            "pregnancy", "recent surgery", "joint problems", "back injury"
        ]
        
        for condition in health_conditions:
            if any(risk in condition.lower() for risk in high_risk_conditions):
                result["concerns"].append(f"Health condition '{condition}' requires professional guidance")
                result["recommendations"].append("Medical clearance strongly recommended before starting program")
                
       
        if goal == Goal.GENTLE_WEIGHT_LOSS:
            result["recommendations"].append("Sustainable weight loss is typically 0.5-2 pounds per week")
            
        return result

def log_health_recommendation(user_id: str, recommendation_type: str, safety_check: Dict[str, Any]):
    """Log health recommendations for safety tracking"""
    logger.info(f"Health recommendation generated for user {user_id}")
    logger.info(f"Type: {recommendation_type}")
    logger.info(f"Safety validation: {safety_check}")
    
    if not safety_check.get("is_valid", True) or safety_check.get("warnings"):
        logger.warning(f"Safety concerns for user {user_id}: {safety_check}")

File: nodeJs-user-service/utils/healthSafety.js
import * as enums from '@/constants/enums'

export class HealthSafetyValidator {
    
    static validateUserProfileSafety(profileData) {
        const result = {
            is_safe: true,
            concerns: [],
            recommendations: [],
            risk_level: enums.RISK_LEVELS.LOW
        }

        const { age, primary_goal, health_conditions = [], current_activity_level } = profileData

        if (age && (age < 16 || age > 75)) {
            result.concerns.push('Age requires special consideration for health planning')
            result.recommendations.push('Professional medical supervision strongly recommended')
            result.risk_level = enums.RISK_LEVELS.MODERATE
        }

        const highRiskConditions = [
            'heart disease', 'cardiovascular', 'cardiac', 'diabetes', 'diabetic',
            'high blood pressure', 'hypertension', 'eating disorder', 'anorexia', 'bulimia',
            'pregnancy', 'pregnant', 'recent surgery', 'surgery', 'joint problems',
            'arthritis', 'back injury', 'spine', 'chest pain', 'shortness of breath',
            'dizziness', 'fainting', 'seizure', 'medication', 'chronic pain'
        ]

        for (const condition of health_conditions) {
            if (typeof condition === 'string') {
                const conditionLower = condition.toLowerCase()
                const hasHighRisk = highRiskConditions.some(risk => 
                    conditionLower.includes(risk)
                )
                
                if (hasHighRisk) {
                    result.concerns.push(`Health condition "${condition}" requires professional medical evaluation`)
                    result.recommendations.push('Medical clearance required before starting any fitness program')
                    result.risk_level = enums.RISK_LEVELS.HIGH
                    result.is_safe = false
                }
            }
        }

        if (primary_goal === enums.HEALTH_GOALS.WEIGHT_LOSS) {
            result.recommendations.push('Weight loss goals should be supervised by healthcare professionals')
            result.recommendations.push('Sustainable weight loss is typically 0.5-2 pounds per week')
        }

        if (primary_goal === enums.HEALTH_GOALS.INJURY_RECOVERY) {
            result.concerns.push('Injury recovery requires professional physical therapy guidance')
            result.recommendations.push('Consult with physical therapist or sports medicine physician')
            result.risk_level = enums.RISK_LEVELS.HIGH
        }
        if (current_activity_level === enums.ACTIVITY_LEVELS.SEDENTARY) {
            result.recommendations.push('Gradual increase in activity recommended for sedentary individuals')
            result.recommendations.push('Start with low-intensity activities and build slowly')
        }

       
        const mentalHealthGoals = [enums.HEALTH_GOALS.STRESS_REDUCTION]
        if (mentalHealthGoals.includes(primary_goal)) {
            result.recommendations.push('Mental health goals benefit from professional counseling support')
        }

        if (result.risk_level === enums.RISK_LEVELS.HIGH || result.concerns.length > 2) {
            result.is_safe = false
        }

        return result
    }

    static validateCalorieTarget(calories, age, goal) {
        const result = {
            is_valid: true,
            adjusted_calories: calories,
            warnings: [],
            recommendations: []
        }

       
        if (calories < enums.SAFETY_LIMITS.MIN_CALORIES) {
            result.is_valid = false
            result.adjusted_calories = enums.SAFETY_LIMITS.MIN_CALORIES
            result.warnings.push(`Calorie target too low. Adjusted to safe minimum: ${enums.SAFETY_LIMITS.MIN_CALORIES}`)
            result.recommendations.push('Very low calorie diets require medical supervision')
        }

        if (calories > enums.SAFETY_LIMITS.MAX_CALORIES) {
            result.is_valid = false
            result.adjusted_calories = enums.SAFETY_LIMITS.MAX_CALORIES
            result.warnings.push(`Calorie target too high. Adjusted to safe maximum: ${enums.SAFETY_LIMITS.MAX_CALORIES}`)
        }

        if (age && age > 65 && calories < 1500) {
            result.warnings.push('Older adults require adequate nutrition - very low calories not recommended')
        }

        if (age && age < 18) {
            result.warnings.push('Growing individuals require adequate calories for healthy development')
            result.recommendations.push('Pediatric nutrition specialist consultation recommended')
        }

        
        if (goal === enums.HEALTH_GOALS.WEIGHT_LOSS && calories < 1500) {
            result.warnings.push('Extreme calorie restriction can be dangerous and unsustainable')
            result.recommendations.push('Consult registered dietitian for safe weight loss approach')
        }

        return result
    }

    static validateWorkoutPlan(workoutMinutesPerDay, activityLevel, age) {
        const result = {
            is_valid: true,
            adjusted_minutes: workoutMinutesPerDay,
            warnings: [],
            recommendations: []
        }

      
        if (workoutMinutesPerDay < enums.SAFETY_LIMITS.MIN_WORKOUT_MINUTES) {
            result.adjusted_minutes = enums.SAFETY_LIMITS.MIN_WORKOUT_MINUTES
            result.recommendations.push('Even short workouts provide health benefits')
        }

        if (workoutMinutesPerDay > enums.SAFETY_LIMITS.MAX_WORKOUT_MINUTES) {
            result.is_valid = false
            result.adjusted_minutes = enums.SAFETY_LIMITS.MAX_WORKOUT_MINUTES
            result.warnings.push('Workout duration exceeds recommended limits for general population')
            result.recommendations.push('Extended exercise requires professional supervision')
        }

        if (activityLevel === enums.ACTIVITY_LEVELS.SEDENTARY && workoutMinutesPerDay > 45) {
            result.warnings.push('Gradual increase in activity recommended for sedentary individuals')
            result.recommendations.push('Start with 15-20 minutes and increase gradually each week')
        }

        
        if (age && age > 65 && workoutMinutesPerDay > 60) {
            result.warnings.push('Adults over 65 should consult healthcare providers for exercise guidance')
            result.recommendations.push('Focus on balance, strength, and cardiovascular health')
        }

        if (age && age < 18) {
            result.recommendations.push('Youth fitness should be age-appropriate and fun')
            result.recommendations.push('Avoid excessive training that could affect growth')
        }

        return result
    }

    static detectEmergencySymptoms(symptoms) {
        const emergencyKeywords = [
            'chest pain', 'severe chest pain', 'heart attack',
            'can\'t breathe', 'difficulty breathing', 'shortness of breath',
            'severe dizziness', 'fainting', 'passed out', 'unconscious',
            'severe nausea', 'vomiting blood', 'blood vomit',
            'broken bone', 'fracture', 'severe injury',
            'allergic reaction', 'anaphylaxis', 'swelling throat',
            'suicidal', 'self harm', 'want to die'
        ]

        const emergencyFlags = []
        
        for (const symptom of symptoms) {
            if (typeof symptom === 'string') {
                const symptomLower = symptom.toLowerCase()
                const hasEmergency = emergencyKeywords.some(keyword => 
                    symptomLower.includes(keyword)
                )
                
                if (hasEmergency) {
                    emergencyFlags.push({
                        symptom,
                        severity: 'critical',
                        requires_immediate_attention: true,
                        emergency_services_recommended: true
                    })
                }
            }
        }

        return {
            has_emergency_symptoms: emergencyFlags.length > 0,
            emergency_flags: emergencyFlags,
            immediate_action_required: emergencyFlags.length > 0
        }
    }

    static validateDietaryRestrictions(restrictions, goal) {
        const result = {
            is_valid: true,
            warnings: [],
            recommendations: []
        }

        if (restrictions.length > 3) {
            result.warnings.push('Multiple dietary restrictions may require professional nutrition guidance')
            result.recommendations.push('Registered dietitian consultation recommended for complex restrictions')
        }
        if (restrictions.includes(enums.DIETARY_RESTRICTIONS.VEGAN) && 
            goal === enums.HEALTH_GOALS.MUSCLE_GAIN) {
            result.recommendations.push('Vegan muscle building requires careful protein planning')
            result.recommendations.push('Consider plant-based protein supplements and variety')
        }

        const medicalRestrictions = [
            enums.DIETARY_RESTRICTIONS.DIABETIC_FRIENDLY,
            enums.DIETARY_RESTRICTIONS.HEART_HEALTHY,
            enums.DIETARY_RESTRICTIONS.LOW_SODIUM
        ]

        const hasMedicalRestriction = restrictions.some(restriction => 
            medicalRestrictions.includes(restriction)
        )

        if (hasMedicalRestriction) {
            result.recommendations.push('Medical dietary restrictions should be reviewed with healthcare provider')
            result.warnings.push('Monitor health parameters when following medical diet plans')
        }

        return result
    }

    static getSafetyDisclaimer() {
        return enums.HEALTH_DISCLAIMER
    }

    static getExerciseDisclaimer() {
        return enums.EXERCISE_DISCLAIMER
    }

    static getNutritionDisclaimer() {
        return enums.NUTRITION_DISCLAIMER
    }
}


export const logHealthRecommendation = (userId, recommendationType, safetyCheck) => {
    const logEntry = {
        timestamp: new Date().toISOString(),
        userId: userId.toString(),
        recommendationType,
        safetyValidation: safetyCheck,
        serviceVersion: '1.0.0'
    }
    
    
    console.log(`[WELLNESS-HEALTH-AUDIT] ${JSON.stringify(logEntry)}`)
    
    return logEntry
}